File: src/main/java/com/github/pareronia/i18n_puzzles/Puzzle2025_12.java
package com.github.pareronia.i18n_puzzles;

import static java.util.stream.Collectors.joining;

import java.text.Collator;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import com.github.pareronia.i18n_puzzles.common.Sample;
import com.github.pareronia.i18n_puzzles.common.Samples;
import com.github.pareronia.i18n_puzzles.common.SolutionBase;

public class Puzzle2025_12 extends SolutionBase<Long>{
    private static final Locale ENGLISH = Locale.ENGLISH;
    private static final Locale SWEDISH = Locale.of("sv");
    private static final Locale DUTCH = Locale.of("nl", "NL");

    private Puzzle2025_12(final boolean debug) {
        super(debug);
    }

    public static Puzzle2025_12 create() {
        return new Puzzle2025_12(false);
    }

    public static Puzzle2025_12 createDebug() {
        return new Puzzle2025_12(true);
    }

    @Override
    public Long solve(final List<String> input) {
        final List<Entry> entries
                = input.stream().map(Entry::fromInput).toList();
        return Set.of(ENGLISH, SWEDISH, DUTCH).stream()
            .map(locale -> entries.stream()
                    .sorted(Entry.comparator(locale))
                    .skip(entries.size() / 2)
            .findFirst().orElseThrow())
            .mapToLong(Entry::phone)
            .reduce(1L, (a, b) -> a * b);
    }

    @Override
    @Samples({
        @Sample(input = TEST, expected = "1885816494308838")
    })
    public void samples() { }

    public static void main(final String[] args) throws Exception {
        Puzzle2025_12.createDebug().run();
    }

    private static final String TEST = """
            Ñíguez Peña, María de los Ángeles: 0151605
            Åberg, Rosa-Maria: 0110966
            Özaydın, Zeynep: 0185292
            van den Heyden, Harm: 0168131
            Ämtler, Lorena: 0112717
            Olofsson, Mikael: 0103652
            van Leeuwen, Joke: 0172199
            Vandersteen, Willy: 0120659
            Østergård, Magnus: 0113959
            van Leeuw, Floor: 0144158
            Navarrete Ortiz, Dolores: 0119411
            Aalto, Alvar: 0192872
            Zondervan, Jan Peter: 0103008
            Æbelø, Aurora: 0113267
            O'Neill, Cara: 0109551
            """;

    record Entry(String last, String infixes, String first, long phone) {
        public static Entry fromInput(final String input) {
            final String[] split = input.split(": ");
            final String[] nameSplit = split[0].split(", ");
            final String infixes = Arrays.stream(nameSplit[0].split(" "))
                .takeWhile(sp -> Character.isLowerCase(sp.charAt(0)))
                .collect(joining(" "));
            final String last = Arrays.stream(nameSplit[0].split(" "))
                .dropWhile(sp -> Character.isLowerCase(sp.charAt(0)))
                .collect(joining(" "));
            return new Entry(
                    last, infixes, nameSplit[1], Long.parseLong(split[1]));
        }

        private static String n(final String s, final Locale locale) {
            final String t = s.replaceAll("\\s+", "").replaceAll("\\'", "")
                .replace('ı', 'i').toUpperCase();
            if (locale == SWEDISH) {
                return t.replace('Ø', 'Ö').replace('Æ', 'Ä');
            }
            return t.replace(String.valueOf('Æ'), "ae").replace('Ø', 'O');
        }
        
        private String lastName(final Locale locale) {
            return switch(locale) {
                case final Locale l
                        when l.equals(DUTCH) -> n(last, l) + n(infixes, l);
                case final Locale l -> n(infixes, l) + n(last, l);
            };
        }

        public static Comparator<Entry> comparator(final Locale locale) {
            final Collator collator = Collator.getInstance(locale);
            final Comparator<Entry> byLastName = (o1, o2)
                -> collator.compare(o1.lastName(locale), o2.lastName(locale));
            final Comparator<Entry> byFirstName = (o1, o2)
                -> collator.compare(n(o1.first, locale), n(o2.first, locale));
            return byLastName.thenComparing(byFirstName);
        }
    }
}

File: src/main/python/Puzzle2025_12.py
from __future__ import annotations

import contextlib
import itertools
import locale
import sys
from math import prod
from typing import Any
from typing import Generator
from typing import NamedTuple

from i18n_puzzles.common import InputData
from i18n_puzzles.common import SolutionBase
from i18n_puzzles.common import i18n_puzzles_samples

TEST = """\
Ñíguez Peña, María de los Ángeles: 0151605
Åberg, Rosa-Maria: 0110966
Özaydın, Zeynep: 0185292
van den Heyden, Harm: 0168131
Ämtler, Lorena: 0112717
Olofsson, Mikael: 0103652
van Leeuwen, Joke: 0172199
Vandersteen, Willy: 0120659
Østergård, Magnus: 0113959
van Leeuw, Floor: 0144158
Navarrete Ortiz, Dolores: 0119411
Aalto, Alvar: 0192872
Zondervan, Jan Peter: 0103008
Æbelø, Aurora: 0113267
O'Neill, Cara: 0109551
"""

Output = int
DUTCH = "nl_NL"
ENGLISH = "en"
SWEDISH = "sv"


@contextlib.contextmanager
def temp_locale(new_locale: str) -> Generator[Any]:
    original_locale = locale.setlocale(locale.LC_ALL)
    try:
        yield locale.setlocale(locale.LC_ALL, new_locale)
    finally:
        locale.setlocale(locale.LC_ALL, original_locale)


class Entry(NamedTuple):
    last: str
    infixes: str
    first: str
    phone: int

    @classmethod
    def from_input(cls, line: str) -> Entry:
        person, phone = line.split(": ")
        last_name, first_name = person.split(", ")
        infixes = "".join(
            itertools.takewhile(lambda ch: ch.islower(), last_name.split())
        )
        last = "".join(
            itertools.dropwhile(lambda ch: ch.islower(), last_name.split())
        )
        return cls(
            last.replace("'", "").replace(" ", ""),
            infixes.replace("'", "").replace(" ", ""),
            first_name.replace("'", "").replace(" ", ""),
            int(phone),
        )

    def collation_key(self, loc: str) -> str:
        if loc == DUTCH:
            return self.last
        elif loc == SWEDISH:
            return (self.infixes + self.last).replace("Æ", "Å")
        else:
            return self.infixes + self.last


class Solution(SolutionBase[Output]):
    def solve(self, input: InputData) -> Output:
        def sort(loc: str) -> list[Entry]:
            with temp_locale(loc):
                return sorted(
                    entries, key=lambda e: locale.strxfrm(e.collation_key(loc))
                )

        entries = [Entry.from_input(line) for line in input]
        return prod(
            sort(loc)[len(entries) // 2].phone
            for loc in {ENGLISH, DUTCH, SWEDISH}
        )

    @i18n_puzzles_samples(((TEST, 1885816494308838),))
    def samples(self) -> None:
        pass


solution = Solution(12)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
